syntax = "proto3";

service RaftService {
    rpc AppendEntries (AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc RequestVote (RequestVoteRequest) returns (RequestVoteResponse);
    rpc Replicate (ReplicateEntriesRequest) returns (ReplicateEntriesResponse);
}

enum Command {
    PUT = 0;
    GET = 1;
}

enum NodeState {
    LEADER = 0;
    FOLLOWER = 1;
    CANDIDATE = 2;
}

message LogEntry {
    uint32 term = 1;
    uint32 index = 2;
    bytes payload = 4;
}

message AppendEntriesRequest {
    uint32 term = 1;
    string leaderId = 2;
    uint32 prevLogIndex = 3;
    uint32 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    uint32 leaderCommit = 6;
    uint32 senderId = 7;
}

message AppendEntriesResponse {
    uint32 term = 1;
    bool success = 2;
    uint32 match_index = 3;
    uint32 responderId = 4;
}

message RequestVoteRequest {
    uint32 term = 1;
    uint32 candidateId = 2;
    uint32 lastLogIndex = 3;
    uint32 lastLogterm = 4;
    uint32 senderId = 7;
}

message RequestVoteResponse {
    uint32 term = 1;
    uint32 voteGranted = 2;
    uint32 responderId = 3;
}

message ReplicateEntriesRequest {
    uint32 term = 1;
    uint32 senderId = 2;
    repeated string payloads = 3;
}

message ReplicateEntriesResponse {
    string status = 1;
}